### ISSUE TITLE
callable type guard narrows too aggressively

### ISSUE THREAD

**Issue by robert-bn** [microsoft/pyright#8944]
Consider the following code:

```python
def p(x: type[int] | object):
    if callable(x):
        reveal_type(x)  # Type of "x" is "type[int]"
```

Pyright narrows the type to of x to `type[int]`, but I would expect it to be `type[int] | <callable subtype of object>`.

This can be used to produce a crash at runtime that pyright finds no issue with.
For example:
```python
from typing import assert_never

def p(x: type[int] | object):
    if callable(x):
        if x is str:
            assert_never(x)
            
p(str)
```

There is another potentially related issue, consider this:
```python
def q(x: object):
    if callable(x) and x is str:
        assert_never(x)
```
Clearly line 3 is reachable if `q` is called with `str`, but pyright finds no issue with this.

Tested with pyright 1.1.379 CLI.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.380.

**Comment by robert-bn** [comment]
@erictraut 

Thanks for addressing this!

It seems like the following code still has an issue in 1.1.380

```python
from typing import assert_type

def example(x: object):
    if callable(x) and x is str:
        assert_type(x, type[str])
```

pyright has the following output
```python
error: "assert_type" mismatch: expected "type[str]" but received "(...) -> object" 
```
I suppose `(...) -> object` is a superclass of `str`, but I would have expected the narrowed type to be `type[str]`.

Should I make a new issue for this?
